### PR TITLE
feat: cache argo cd rbac

### PR DESCRIPTION
**Pull request by pasha-codefresh**
Continuation on https://github.com/argoproj/argo-cd/pull/6739/files

### CODE REVIEW

**Comment by yeya24** [comment]
IIUC, the `AddFunction` call needs to invalidate the cache as well?

**Comment by yeya24** [comment]
We also implemented the rbac cache based on #6739 and have been using it on prod for several weeks. The issue of the casbin cached enforcer is that it doesn't free memory and will hold cached items forever.

**Comment by pasha-codefresh** [comment]
Thank you for comment @yeya24

File: util/rbac/rbac.go [hunk 1]
```diff
@@ -181,28 +219,28 @@ func (e *Enforcer) EnforceRuntimePolicy(policy string, rvals ...interface{}) boo
...
-func (e *Enforcer) CreateEnforcerWithRuntimePolicy(policy string) *casbin.Enforcer {
-	var enf *casbin.Enforcer
+func (e *Enforcer) CreateEnforcerWithRuntimePolicy(policy string) WrapperEnforcer {
+	var enf WrapperEnforcer
 	var err error
 	if policy == "" {
-		enf = e.Enforcer
+		enf = e.wrapperEnforcer
 	} else {
 		enf, err = newEnforcerSafe(newBuiltInModel(), newAdapter(e.adapter.builtinPolicy, e.adapter.userDefinedPolicy, policy))
```
Comment (alexmt) [inline util/rbac/rbac.go:228]: If the project has roles with policies then the cache is not used. We should support this use case as well and use cached enforces with project with policies.

**Review by alexmt — CHANGES_REQUESTED** [review]

**Review by alexmt — APPROVED** [review]
LGTM

**Comment by alexmt** [comment]
Thanks a lot to @farodin91 for working on initial implementation in  https://github.com/argoproj/argo-cd/pull/6739!

**Pull request merged**